### ISSUE TITLE
Frequency weighted training sentences

### ISSUE THREAD

**Issue by feddyfedfed** [google/sentencepiece#181]
I don't know if this is already implemented or if there's a workaround for this.  Sometimes, due to the large amount of data, what we have are training sentences that are already uniquely condensed (and we just keep their frequency counts somewhere).  Of course, this condensed training set does not necessarily reflect the original statistics of the occurrences of each word and may not reflect the optimal choices for the tokens.

Through experiments, I create two separate segmentation models using the (1) 10M most frequent sentences (unique), and (2) 10M random sentences from the entire set (duplicates are spread in the chronological order that they were generated).   Case (2) is of course favored by the empirical evidence suggested by the developers.  However, for our task I found that case (1) model is performing better.  So the experiment was inconclusive in deciding on which seeds set to choose next time.  

I also found that the maximum number of training set is 100M and I wanted to take advantage of this limit since we have a huge amount of data.  Of course I could simply generate random 100M sentences from all the training data, but I found it quite empirical.  Especially since we're working on CJK languages.

So I guess in a nutshell:
(1) Is there a way to incorporate frequency counts of sentences?
(2) What is a statistically reliable way to condense a huge amount of data (let's say the original size is 300% of the maximum allowed), but still reflect the (approximate) statistics of the words?

**Comment by taku910** [comment]
Thank you for using SentencePiece.

Though it is not documented, `spm_train --input=tsv_file --input_format=tsv ..`  allows for loading a corpus in tsv format where the first column is the sentence and the second column is frequency.  `--input_format=tsv` is originally prepared to train the model from the word list with frequency.

As for the training data, I feel that the default size 100M is enough for SPM training as long as they are sampled appropriately from the full corpus.  SentencePiece adopts a simple uni-gram language model with relatively smaller vocab size.  We would be able to get a reliable statistics only from the small subset of the corpus.  However, it would be nicer to check the output segmentation changing the size of corpus.

**Comment by feddyfedfed** [comment]
Thank you Kudo-san for the satisfying answers and thank you for continuously developing SentencePiece!  Cheers.
